### PR TITLE
[SPARK-47207][CORE] Support `spark.driver.timeout` and `DriverTimeoutPlugin`

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
### What changes were proposed in this pull request?

This PR aims to support `spark.driver.timeout` and `DriverTimeoutPlugin`.

### Why are the changes needed?

Sometime, Spark applications fall into abnormal situation and hang.

We had better provide a way to guarantee the termination after pre-defined timeout via a standard way.
- spark.plugins=org.apache.spark.deploy.DriverTimeoutPlugin
- spark.driver.timeout=1min

```
$ bin/spark-shell -c spark.plugins=org.apache.spark.deploy.DriverTimeoutPlugin -c spark.driver.timeout=1min
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 4.0.0-SNAPSHOT
      /_/

Using Scala version 2.13.12 (OpenJDK 64-Bit Server VM, Java 17.0.10)
Type in expressions to have them evaluated.
Type :help for more information.
24/02/28 06:53:34 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Spark context Web UI available at http://localhost:4040
Spark context available as 'sc' (master = local[*], app id = local-1709132014477).
Spark session available as 'spark'.

scala> 24/02/28 06:54:34 WARN DriverTimeoutDriverPlugin: Terminate Driver JVM because it runs after 1 minute

$ echo $?
124
```

### Does this PR introduce _any_ user-facing change?

No, this is a new feature and a built-in plugin.

### How was this patch tested?

Manually because this invokes `System.exit`.

1. Timeout with 1 minute
```
$ bin/spark-shell -c spark.plugins=org.apache.spark.deploy.DriverTimeoutPlugin -c spark.driver.timeout=1min
...
scala> 24/02/28 06:54:34 WARN DriverTimeoutDriverPlugin: Terminate Driver JVM because it runs after 1 minute

$ echo $?
124
```

2. `DriverTimeoutPlugin` will be ignored if the default value of `spark.driver.timeout` is used.
```
$ bin/spark-shell -c spark.plugins=org.apache.spark.deploy.DriverTimeoutPlugin
...
24/02/28 01:02:57 WARN DriverTimeoutDriverPlugin: Disabled with the timeout value 0.
...
scala>
```

3. `spark.driver.timeout` will be ignored if `DriverTimeoutPlugin` is not provided.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Could you review this PR too, @LuciferYang ?

File: core/src/main/scala/org/apache/spark/internal/config/package.scala [hunk 1]
```diff
@@ -1099,6 +1099,15 @@ package object config {
     .booleanConf
     .createWithDefault(false)
 
+  private[spark] val DRIVER_TIMEOUT = ConfigBuilder("spark.driver.timeout")
```
[the remark this answers is not in the document]
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/internal/config/package.scala:1102]: Sure.

**Comment by dongjoon-hyun** [comment]
The PR is updated according to your advice, @LuciferYang .

**Review by LuciferYang — APPROVED** [review]
LGTM

**Comment by dongjoon-hyun** [comment]
Thank you so much! Merged to master for Apache Spark 4.0.0. This plugin PR is irrelevant to CIs.

**Pull request closed without merging**